### ISSUE TITLE
Conditional Access Policy: Resultant PS1 can't compile to MOF

### ISSUE THREAD

**Issue by mlhickey** [microsoft/Microsoft365DSC#2184]
Exported CA policies can't compile into MOF due to Test-ConflictingResources conflict.

#### Suggested solution to the issue
Use Id instead of DisplayName as key as duplicate displaynames are allowed.
#### The DSC configuration that is used to reproduce the issue (as detailed as possible)
```powershell
# Generated with Microsoft365DSC version 1.22.720.1
# For additional information on how to use Microsoft365DSC, please visit https://aka.ms/M365DSC
Configuration ppe
{
    $OrganizationName = $ConfigurationData.NonNodeData.OrganizationName
    Import-DscResource -ModuleName 'Microsoft365DSC' -ModuleVersion '1.22.720.1'

    Node localhost
    {

        AADConditionalAccessPolicy 81d8765a-8fa6-4806-bf1d-d1eeda6ce495
        {
            ApplicationEnforcedRestrictionsIsEnabled = $False;
            ApplicationId                            = $ConfigurationData.NonNodeData.ApplicationId;
            BuiltInControls                          = @("mfa");
            CertificateThumbprint                    = $ConfigurationData.NonNodeData.CertificateThumbprint;
            ClientAppTypes                           = @("browser","mobileAppsAndDesktopClients");
            CloudAppSecurityIsEnabled                = $False;
            CloudAppSecurityType                     = "";
            DeviceFilterRule                         = "";
            DisplayName                              = "100 - Admin protection - All apps: Require MFA For admins";
            Ensure                                   = "Present";
            ExcludeApplications                      = @();
            ExcludeDevices                           = @();
            ExcludeGroups                            = @();
            ExcludeLocations                         = @();
            ExcludePlatforms                         = @();
            ExcludeRoles                             = @();
            ExcludeUsers                             = @();
            GrantControlOperator                     = "OR";
            Id                                       = "4004cfd9-4dfe-47a7-b721-1579020235e6";
            IncludeApplications                      = @("All");
            IncludeDevices                           = @();
            IncludeGroups                            = @();
            IncludeLocations                         = @();
            IncludePlatforms                         = @();
            IncludeRoles                             = @("Application Administrator","Application Developer","Attack Payload Author","Attack Simulation Administrator","Attribute Assignment Administrator","Attribute Assignment Reader","Attribute Definition Administrator","Attribute Definition Reader","Authentication Administrator","Authentication Policy Administrator","Azure DevOps Administrator","Azure Information Protection Administrator","Azure AD Joined Device Local Administrator","B2C IEF Policy Administrator","B2C IEF Keyset Administrator","Billing Administrator","Cloud App Security Administrator","Cloud Application Administrator","Cloud Device Administrator","Compliance Administrator","Compliance Data Administrator","Conditional Access Administrator","Customer LockBox Access Approver","Desktop Analytics Administrator","Directory Readers","Directory Synchronization Accounts","Directory Writers","Domain Name Administrator","Edge Administrator","Exchange Administrator","Dynamics 365 Administrator","Exchange Recipient Administrator","External ID User Flow Administrator","External ID User Flow Attribute Administrator","External Identity Provider Administrator","Global Administrator","Global Reader","Groups Administrator","Guest Inviter","Helpdesk Administrator","Hybrid Identity Administrator","Identity Governance Administrator","Insights Administrator","Insights Business Leader","Intune Administrator","Kaizala Administrator","Knowledge Administrator","Knowledge Manager","License Administrator","Message Center Privacy Reader","Message Center Reader","Network Administrator","Office Apps Administrator","Password Administrator","Power BI Administrator","Power Platform Administrator","Printer Administrator","Privileged Authentication Administrator","Printer Technician","Privileged Role Administrator","Search Administrator","Reports Reader","Search Editor","Security Administrator","Security Operator","Security Reader","Service Support Administrator","SharePoint Administrator","Skype for Business Administrator","Teams Administrator","Teams Communications Administrator","Teams Communications Support Engineer","Teams Communications Support Specialist","Teams Devices Administrator","Usage Summary Reports Reader","User Administrator","Windows 365 Administrator","Windows Update Deployment Administrator");
            IncludeUserActions                       = @();
            IncludeUsers                             = @();
            PersistentBrowserIsEnabled               = $False;
            PersistentBrowserMode                    = "";
            SignInFrequencyIsEnabled                 = $False;
            SignInFrequencyType                      = "";
            SignInRiskLevels                         = @();
            State                                    = "enabledForReportingButNotEnforced";
            TenantId                                 = $ConfigurationData.NonNodeData.TenantId;
            UserRiskLevels                           = @();
        }

        AADConditionalAccessPolicy aedaf45d-a7eb-42e0-b18a-a363f149e291
        {
            ApplicationEnforcedRestrictionsIsEnabled = $False;
            ApplicationId                            = $ConfigurationData.NonNodeData.ApplicationId;
            BuiltInControls                          = @("compliantDevice","domainJoinedDevice");
            CertificateThumbprint                    = $ConfigurationData.NonNodeData.CertificateThumbprint;
            ClientAppTypes                           = @("mobileAppsAndDesktopClients");
            CloudAppSecurityIsEnabled                = $False;
            CloudAppSecurityType                     = "";
            DeviceFilterRule                         = "";
            DisplayName                              = "504 - <RING> - Data protection - O365: Require trusted device When using desktop clients on Windows and macOS";
            Ensure                                   = "Present";
            ExcludeApplications                      = @();
            ExcludeDevices                           = @();
            ExcludeGroups                            = @();
            ExcludeLocations                         = @();
            ExcludePlatforms                         = @();
            ExcludeRoles                             = @();
            ExcludeUsers                             = @();
            GrantControlOperator                     = "OR";
            Id                                       = "150ceb71-572a-4ad0-9f63-05583b787418";
            IncludeApplications                      = @("Office365");
            IncludeDevices                           = @();
            IncludeGroups                            = @();
            IncludeLocations                         = @();
            IncludePlatforms                         = @("windows","macOS");
            IncludeRoles                             = @();
            IncludeUserActions                       = @();
            IncludeUsers                             = @("All");
            PersistentBrowserIsEnabled               = $False;
            PersistentBrowserMode                    = "";
            SignInFrequencyIsEnabled                 = $False;
            SignInFrequencyType                      = "";
            SignInRiskLevels                         = @();
            State                                    = "enabledForReportingButNotEnforced";
            TenantId                                 = $ConfigurationData.NonNodeData.TenantId;
            UserRiskLevels                           = @();
        }

        AADConditionalAccessPolicy 1ccc489e-6501-4c2e-8544-29af64c64893
        {
            ApplicationEnforcedRestrictionsIsEnabled = $False;
            ApplicationId                            = $ConfigurationData.NonNodeData.ApplicationId;
            BuiltInControls                          = @("mfa");
            CertificateThumbprint                    = $ConfigurationData.NonNodeData.CertificateThumbprint;
            ClientAppTypes                           = @("browser","mobileAppsAndDesktopClients");
            CloudAppSecurityIsEnabled                = $False;
            CloudAppSecurityType                     = "";
            DeviceFilterRule                         = "";
            DisplayName                              = "100 - Admin protection - All apps: Require MFA For admins";
            Ensure                                   = "Present";
            ExcludeApplications                      = @();
            ExcludeDevices                           = @();
            ExcludeGroups                            = @();
            ExcludeLocations                         = @();
            ExcludePlatforms                         = @();
            ExcludeRoles                             = @();
            ExcludeUsers                             = @();
            GrantControlOperator                     = "OR";
            Id                                       = "3cc76a58-cbc6-46ac-9572-b4705778a0ac";
            IncludeApplications                      = @("All");
            IncludeDevices                           = @();
            IncludeGroups                            = @();
            IncludeLocations                         = @();
            IncludePlatforms                         = @();
            IncludeRoles                             = @("Application Administrator","Application Developer","Attack Payload Author","Attack Simulation Administrator","Attribute Assignment Administrator","Attribute Assignment Reader","Attribute Definition Administrator","Attribute Definition Reader","Authentication Administrator","Authentication Policy Administrator","Azure DevOps Administrator","Azure Information Protection Administrator","Azure AD Joined Device Local Administrator","B2C IEF Policy Administrator","B2C IEF Keyset Administrator","Billing Administrator","Cloud App Security Administrator","Cloud Application Administrator","Cloud Device Administrator","Compliance Administrator","Compliance Data Administrator","Conditional Access Administrator","Customer LockBox Access Approver","Desktop Analytics Administrator","Directory Readers","Directory Synchronization Accounts","Directory Writers","Domain Name Administrator","Edge Administrator","Exchange Administrator","Dynamics 365 Administrator","Exchange Recipient Administrator","External ID User Flow Administrator","External ID User Flow Attribute Administrator","External Identity Provider Administrator","Global Administrator","Global Reader","Groups Administrator","Guest Inviter","Helpdesk Administrator","Hybrid Identity Administrator","Identity Governance Administrator","Insights Administrator","Insights Business Leader","Intune Administrator","Kaizala Administrator","Knowledge Administrator","Knowledge Manager","License Administrator","Message Center Privacy Reader","Message Center Reader","Network Administrator","Office Apps Administrator","Password Administrator","Power BI Administrator","Power Platform Administrator","Printer Administrator","Privileged Authentication Administrator","Printer Technician","Privileged Role Administrator","Search Administrator","Reports Reader","Search Editor","Security Administrator","Security Operator","Security Reader","Service Support Administrator","SharePoint Administrator","Skype for Business Administrator","Teams Administrator","Teams Communications Administrator","Teams Communications Support Engineer","Teams Communications Support Specialist","Teams Devices Administrator","Usage Summary Reports Reader","User Administrator","Windows 365 Administrator","Windows Update Deployment Administrator");
            IncludeUserActions                       = @();
            IncludeUsers                             = @();
            PersistentBrowserIsEnabled               = $False;
            PersistentBrowserMode                    = "";
            SignInFrequencyIsEnabled                 = $False;
            SignInFrequencyType                      = "";
            SignInRiskLevels                         = @();
            State                                    = "enabledForReportingButNotEnforced";
            TenantId                                 = $ConfigurationData.NonNodeData.TenantId;
            UserRiskLevels                           = @();
        }
        AADConditionalAccessPolicy 4a4e6472-0add-4022-bfb6-354c6b138b99
        {
            ApplicationEnforcedRestrictionsIsEnabled = $False;
            ApplicationId                            = $ConfigurationData.NonNodeData.ApplicationId;
            BuiltInControls                          = @("compliantDevice","domainJoinedDevice");
            CertificateThumbprint                    = $ConfigurationData.NonNodeData.CertificateThumbprint;
            ClientAppTypes                           = @("mobileAppsAndDesktopClients");
            CloudAppSecurityIsEnabled                = $False;
            CloudAppSecurityType                     = "";
            DeviceFilterRule                         = "";
            DisplayName                              = "504 - <RING> - Data protection - O365: Require trusted device When using desktop clients on Windows and macOS";
            Ensure                                   = "Present";
            ExcludeApplications                      = @();
            ExcludeDevices                           = @();
            ExcludeGroups                            = @();
            ExcludeLocations                         = @();
            ExcludePlatforms                         = @();
            ExcludeRoles                             = @();
            ExcludeUsers                             = @();
            GrantControlOperator                     = "OR";
            Id                                       = "ba67d49d-2937-4fa8-a4ff-2b6b4ec43d35";
            IncludeApplications                      = @("Office365");
            IncludeDevices                           = @();
            IncludeGroups                            = @();
            IncludeLocations                         = @();
            IncludePlatforms                         = @("windows","macOS");
            IncludeRoles                             = @();
            IncludeUserActions                       = @();
            IncludeUsers                             = @("All");
            PersistentBrowserIsEnabled               = $False;
            PersistentBrowserMode                    = "";
            SignInFrequencyIsEnabled                 = $False;
            SignInFrequencyType                      = "";
            SignInRiskLevels                         = @();
            State                                    = "enabledForReportingButNotEnforced";
            TenantId                                 = $ConfigurationData.NonNodeData.TenantId;
            UserRiskLevels                           = @();
        }
    }
}
ppe -ConfigurationData .\ConfigurationData.psd1

```

#### The operating system the target node is running

OsName               : Microsoft Windows 11 Pro
OsOperatingSystemSKU : 48
OsArchitecture       : 64-bit
WindowsVersion       : 2009
WindowsBuildLabEx    : 22621.1.amd64fre.ni_release.220506-1250
OsLanguage           : en-US
OsMuiLanguages       : {en-US}
#### Version of the DSC module that was used ('dev' if using current dev branch)
1.22.720.1

**Comment by serial666** [comment]
@mlhickey didn't even notice you suggested the solution before I created the PR, too early in the morning I guess :D

I was able to compile your config to MOF with this.

**Comment by andikrueger** [comment]
Please make sure to use unique Display Names. The display name of the Policy is the key parameter of this resource. 

From a standpoint of usability: Every Conditional Access Policy should use a unique display name that characterizes the settings within this policy. You should be able to create unique names by using a naming convention for that.

**Comment by mlhickey** [comment]
It is very common in enterprise scenarios to use a duplicate name while working through various iterations.  Using the id (which is a system-enforced unique identifier) as a key seems like a better path.  An alternative might be a combo of id & display name.

**Comment by andikrueger** [comment]
I totally agree. The challenge lies in the id. The id is generated during the creation of an object and can’t be controlled by my M365DSC. The requirement of an ID as an (additional) key for a resource would make the processing very complex. Within DSC there are only key and write parameters. Nothing like a combined key or something that could help to work around. DSC verifies the Key parameter for uniqueness across all resources of one type.


Adding the id as additional key would require the key to be set during the first run of the configuration. After the first execution of the configuration this ID would be different and m365dsc would create this resource again…

**Comment by mlhickey** [comment]
👍 Chicken-and-egg ...  ObjectId as key is fine until/unless auto-remediation does a re-create, then it's either drift or re-baseline.  So now it's either use a non-unique identifier as a unique identifier or deal with re-baselining, neither desirable scenarios.

**Comment by andikrueger** [comment]
Tagging the following issues to this one, as they have the same problem and partial solutions for that:

- #1746 - This issue is in regards to teams and got partially fixed in #1749 
- #2006 
- #2006 - This issue mentions the variety of problems with the display name as key and suggests a DisplayNameId param. Discussion ongoing.

I would close this issue here for now and track this issue in #2006. Please join the discussion over there.
